[JSC] Avoid using `MarkedArgumentBuffer` in `createTuple`

sosukesuzuki opened the pull request:
#### 7e9dff0fcda41b9afa1e7d5a6328663e67523b3f
<pre>
[JSC] Avoid using `MarkedArgumentBuffer` in `createTuple`
<a href="https://bugs.webkit.org/show_bug.cgi?id=301078">https://bugs.webkit.org/show_bug.cgi?id=301078</a>

Reviewed by Yusuke Suzuki and Keith Miller.

createTupel creates an array from two JSValues. It has a fixed length
so there is no need to use a MarkedArgumentBuffer.

This patch changes to avoid using `MarkedArgumentBuffer` in `createTuple`.

* Source/JavaScriptCore/runtime/IteratorOperations.h:
(JSC::forEachInMapStorage):
* Source/JavaScriptCore/runtime/JSArray.cpp:
(JSC::constructArrayPair):
* Source/JavaScriptCore/runtime/JSArray.h:
* Source/JavaScriptCore/runtime/JSGlobalObjectInlines.h:
(JSC::createTuple):
* Source/JavaScriptCore/runtime/JSMapIterator.h:
* Source/JavaScriptCore/runtime/JSSetIterator.h:

Canonical link: <a href="https://commits.webkit.org/302067@main">https://commits.webkit.org/302067@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/4caf4619a1b171b44e1e95a736cf14696e869ff2

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/127577 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/47225 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/38375 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/134654 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/79134 "Built successfully") 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/47847 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/55753 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/97117 "Passed tests") | [✅ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/65035 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/130525 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/38269 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/114268 "Passed tests") | [❌ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/77597 "Found 1 new API test failure: WPE/TestWebsiteData:/webkit/WebKitWebsiteData/itp (failure)") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/37082 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/32365 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/78027 "Built successfully") | 
| [✅ 🛠 🧪 jsc](https://ews-build.webkit.org/#/builders/20/builds/119614 "Built successfully and passed tests") | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/108135 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/32806 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/137138 "Built successfully") | 
| [✅ 🛠 🧪 jsc-arm64](https://ews-build.webkit.org/#/builders/12/builds/126040 "Built successfully and passed tests") | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/54236 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/41798 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/105641 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/54747 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/110625 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/105292 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/26924 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/50809 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/29248 "Passed tests") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/51822 "Built successfully") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/54173 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/60260 "Built successfully") | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/35/builds/159078 "Built successfully") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/53407 "Built successfully") | | [✅ 🧪 jsc-armv7-tests](https://ews-build.webkit.org/#/builders/25/builds/39763 "Passed tests") | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/56864 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/55166 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

kmiller68 commented on Source/JavaScriptCore/runtime/JSGlobalObjectInlines.h:
```diff
@@ -384,13 +393,11 @@ ALWAYS_INLINE JSArray* createPatternFilledArray(JSGlobalObject* globalObject, JS
 }
 
 template<typename... Args>
+requires (sizeof...(Args) == 2) && (std::is_convertible_v<Args, JSValue> && ...)
```
Seems like we should get rid of this whole method and just have everyone call `constructArray` directly if we're going to have this requirement.

kmiller68 commented on Source/JavaScriptCore/runtime/JSArray.h:
```diff
@@ -419,6 +419,7 @@ inline bool isJSArray(JSCell* cell)
 
 inline bool isJSArray(JSValue v) { return v.isCell() && isJSArray(v.asCell()); }
 
+JS_EXPORT_PRIVATE JSArray* constructArray(JSGlobalObject*, Structure*, const std::array<JSValue, 2> values);
```
Why use a `std::array`? Why not just do `JSArray* constructArray(JSGlobalObject*, Structure*, const, JSValue first, JSValue second);`?

kmiller68 commented on Source/JavaScriptCore/runtime/JSGlobalObjectInlines.h:
```diff
@@ -295,6 +295,15 @@ inline JSArray* constructEmptyArray(JSGlobalObject* globalObject, ArrayAllocatio
+    Structure* structure = globalObject->arrayStructureForProfileDuringAllocation(globalObject, profile, newTarget);
+    RETURN_IF_EXCEPTION(scope, nullptr);
+    return ArrayAllocationProfile::updateLastAllocationFor(profile, constructArray(globalObject, structure, first, second));
+}
```
Why inline this part but not the construction and initialization?

Anyway, I would also do this as:
```
inline JSArray* constructArrayPair(JSGlobalObject* globalObject, JSValue first, JSValue second)
{
    VM& vm = getVM(globalObject);
    auto scope = DECLARE_THROW_SCOPE(vm);
    IndexingType indexingType = ArrayWithUndecided;
    indexingType = leastUpperBoundOfIndexingTypeAndValue(indexingType, first);
    indexingType = leastUpperBoundOfIndexingTypeAndValue(indexingType, second);
    Structure* structure = globalObject->arrayStructureForIndexingTypeDuringAllocation(indexingType);

    ObjectInitializationScope scope(vm);
    JSArray* array = constructArray<AllocationFailureMode::Assert>(scope, arrayStructure, 2);
    array->initializeIndex(scope, 0, first);
    array->initializeIndex(scope, 1, second);

    return array;
}
```

If we don't have an array allocation profile we should be using the passed values to get the indexing type. Otherwise we'll might end up doing a bunch of indexing type transitions, which are expensive.

Although now that I've written it it might just be cleaner to go back to `createTuple` and have that do a pair of fold expressions for `leastUpperBoundOfIndexingTypeAndValue` and `initializeIndex`.

sosukesuzuki replied:
I created a new function, `constructArrayPair`, instead of `createTuple`. This function needs to be exposed to WebCore with `JS_EXPORT_PRIVAE`. I thought it would be easier to understand if it were placed in the same file with a similar name to the already exposed `constructArray`, and I believe it is sufficiently simple.

kmiller68 reviewed (approved):
r=me, thanks

sosukesuzuki committed:
[JSC] Avoid using `MarkedArgumentBuffer` in `createTuple`
https://bugs.webkit.org/show_bug.cgi?id=301078

Reviewed by Yusuke Suzuki and Keith Miller.

createTupel creates an array from two JSValues. It has a fixed length
so there is no need to use a MarkedArgumentBuffer.

This patch changes to avoid using `MarkedArgumentBuffer` in `createTuple`.

* Source/JavaScriptCore/runtime/IteratorOperations.h:
(JSC::forEachInMapStorage):
* Source/JavaScriptCore/runtime/JSArray.cpp:
(JSC::constructArrayPair):
* Source/JavaScriptCore/runtime/JSArray.h:
* Source/JavaScriptCore/runtime/JSGlobalObjectInlines.h:
(JSC::createTuple):
* Source/JavaScriptCore/runtime/JSMapIterator.h:
* Source/JavaScriptCore/runtime/JSSetIterator.h:

Canonical link: https://commits.webkit.org/302067@main
